File: runtime/cpp/titan-ast-runtime-lib/SingleDerivedTerminalGrammarAutomata.cpp
//
// Created by tian wei jun on 2025/4/6.
//

#include "SingleDerivedTerminalGrammarAutomata.h"

SingleDerivedTerminalGrammarAutomata::SingleDerivedTerminalGrammarAutomata(
    const DerivedTerminalGrammarAutomataData
        *derivedTerminalGrammarAutomataData,
    const TokenDfa *tokenDfa)
    : DfaTokenAutomata(tokenDfa) {
  auto &map =
      derivedTerminalGrammarAutomataData->rootTerminalGrammarMaps.front();
  rootTerminalGrammar = map.rootTerminalGrammar;
  textTerminalMap = &(map.textTerminalMap);
}

// dfa delete by PersistentObject.tokenDfa
// derivedTerminalGrammarAutomataData delete by
// PersistentObject.derivedTerminalGrammarAutomataData tokens delete by caller
SingleDerivedTerminalGrammarAutomata::~SingleDerivedTerminalGrammarAutomata() =
    default;

TokensResult *SingleDerivedTerminalGrammarAutomata::buildToken(
    const std::string *sourceFilePath) {
  auto tokensResult = DfaTokenAutomata::buildToken(sourceFilePath);
  if (tokensResult->isOk()) {
    buildTokenBySingleDerivedTerminalGrammarAutomata(tokensResult->getOkData());
  }
  return tokensResult;
}
std::vector<Token *> *SingleDerivedTerminalGrammarAutomata::
    buildTokenBySingleDerivedTerminalGrammarAutomata(
        std::vector<Token *> *tokens) const {
  for (auto token : *tokens) {
    if (rootTerminalGrammar->index == token->terminal.index) {
      auto terminalIt = textTerminalMap->find(&token->text);
      if (terminalIt != textTerminalMap->end()) {
        auto terminal = terminalIt->second;
        token->terminal = *terminal;
      }
    }
  }
  return tokens;
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/SingleDerivedTerminalGrammarAutomata.java
package titan.ast.runtime;

import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import titan.ast.runtime.AstGeneratorResult.TokensResult;
import titan.ast.runtime.DerivedTerminalGrammarAutomataData.RootTerminalGrammarMap;

/**
 * .
 *
 * @author tian wei jun
 */
class SingleDerivedTerminalGrammarAutomata extends DfaTokenAutomata {
  Grammar rootTerminalGrammar;
  HashMap<String, Grammar> textTerminalMap;

  /**
   * 初始化oneTokenStringBuilder、eof、dfa.
   *
   * @param derivedTerminalGrammarAutomataData 关键字有限自动机
   * @param dfa 识别token的 确定有限状态自动机
   */
  SingleDerivedTerminalGrammarAutomata(
      DerivedTerminalGrammarAutomataData derivedTerminalGrammarAutomataData, TokenDfa dfa) {
    super(dfa);
    RootTerminalGrammarMap rootTerminalGrammarMap =
        derivedTerminalGrammarAutomataData.rootTerminalGrammarMaps.get(0);
    this.rootTerminalGrammar = rootTerminalGrammarMap.rootTerminalGrammar;
    this.textTerminalMap = rootTerminalGrammarMap.textTerminalMap;
  }

  @Override
  public TokensResult buildToken(InputStream byteInputStream) {
    TokensResult tokensResult = super.buildToken(byteInputStream);
    if (tokensResult.isOk()) {
      buildToken(tokensResult.getOkData());
    }
    return tokensResult;
  }

  List<Token> buildToken(List<Token> tokens) {
    for (Token token : tokens) {
      if (token.terminal == rootTerminalGrammar) {
        Grammar terminal = textTerminalMap.get(token.text);
        if (null != terminal) {
          token.terminal = terminal;
        }
      }
    }
    return tokens;
  }
}
